cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
     return String.format("%s:%s:%s:%s", program.getNamespaceId(), program.getApplicationId(),
                          programType.name(), program.getId());
   }
+
+  /**
+   * One instance of this class contains a list of {@link StreamSizeSchedule}s, which are all interested
+   * in the same stream. This instance subscribes to the size notification of the stream, and polls the
+   * stream for its size whenever the schedules it references need the information.
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Object lastNotificationLock;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.lastNotificationLock = new Object();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
```

Comment:
What happens if this fails? This would throw an exception that pops up to some caller trying to create a schedule? Is there a retry logic? And if this succeeds, can it happen that the subscription is interrupted by some failure, will it reconnect?